### PR TITLE
Update vision collection to remove artefacts in collected images

### PR DESCRIPTION
**Pull request by LachlanCourt**
Goal of this PR is to solve issue artefacts cropping up in a number of the images. It works and is ready for testing now I just want to change the config slightly still before we merge. (Gotta go into the lab this arvo to do that but wanted to get this up asap)

TODO
- [x] Remake rotations (I think that they are slightly out of date, it still works though)
- [x] Merge master after Webots migration
- [x] General code tidy up
- [ ] ~Read robots in from file~
- [x] Update README

### CODE REVIEW

File: protos/robot/NUgus.proto
Comment (ysims) [inline protos/robot/NUgus.proto:37]: Revert these changes, I assume you don't use them? This was an early attempt to fix the issue.

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (ysims) [inline controllers/teleport_master_controller/teleport_master_controller.cpp:76]: Should do consistent casing in here, there's a big mix of camelCase and snake_case.

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (ysims) [inline controllers/teleport_master_controller/teleport_master_controller.cpp:72]: Very minor, but `robotNodes` sounds grammatically better?

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (ysims) [inline controllers/teleport_master_controller/teleport_master_controller.cpp:154]:
```suggestion
    // number of timesteps using a modulo operation
```

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (ysims) [inline controllers/teleport_master_controller/teleport_master_controller.cpp:226]: Why not just do inside the while loop
- Move robot
- Call `supervisor.step(timeStep)` (or `timeStep*2` if needed?)
- Save data

Rather than having all the modulo stuff?

Comment (LachlanCourt) [reply controllers/teleport_master_controller/teleport_master_controller.cpp:226]: If that would do the same thing I can change it... It just made more sense in my head to have the while loop run every time rather than stepping halfway through the loop

File: protos/robot/NUgus.proto
Comment (LachlanCourt) [reply protos/robot/NUgus.proto:37]: I don't even remember changing this lmao I'll change it back

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (ysims) [reply controllers/teleport_master_controller/teleport_master_controller.cpp:226]: Up to you, they're just different ways to do it - I thought it might be a bit neater but it doesn't really matter.

File: protos/robot/NUgus.proto
Comment (ysims) [reply protos/robot/NUgus.proto:37]: This was my change, it was an early attempt by me to fix the weird images.

**Comment by ysims** [comment]
Oh also change the PR title so it's not just the branch name.

File: controllers/teleport_master_controller/config.yaml
Comment (KipHamiltons) [inline controllers/teleport_master_controller/config.yaml:45]: What is this? min distance between what? comment please
```suggestion
min_distance: 0.5
```

File: controllers/teleport_master_controller/config.yaml
Comment (KipHamiltons) [inline controllers/teleport_master_controller/config.yaml:46]:
```suggestion
z_height: 0.51
```

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (KipHamiltons) [inline controllers/teleport_master_controller/teleport_master_controller.cpp:44]: I think this is actually padding it with ascii 0, instead of null, which I am pretty sure should be `'\0'`.

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (KipHamiltons) [inline controllers/teleport_master_controller/teleport_master_controller.cpp:49]: If it can be constexpr, it probably should be. It should probably be moved into main too. globals bad
```suggestion
constexpr int QUALITY = 100;
```

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (KipHamiltons) [inline controllers/teleport_master_controller/teleport_master_controller.cpp:61]:
```suggestion
    const std::string fieldDef = argv[1];
```

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (KipHamiltons) [inline controllers/teleport_master_controller/teleport_master_controller.cpp:63]:
```suggestion
    const std::string selfDef = argv[2];
```

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (KipHamiltons) [inline controllers/teleport_master_controller/teleport_master_controller.cpp:69]: can this be const too?

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (KipHamiltons) [inline controllers/teleport_master_controller/teleport_master_controller.cpp:76]:
```suggestion
    const int timeStep = int(supervisor.getBasicTimeStep());
```

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (KipHamiltons) [inline controllers/teleport_master_controller/teleport_master_controller.cpp:80]:
```suggestion
    const webots::Camera* leftCamera  = supervisor.getCamera("left_camera");
    const webots::Camera* rightCamera = supervisor.getCamera("right_camera");
```

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (KipHamiltons) [inline controllers/teleport_master_controller/teleport_master_controller.cpp:90]:
```suggestion
    const int cameraWidth             = leftCamera->getWidth();
    const int cameraHeight            = leftCamera->getHeight();
```

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (KipHamiltons) [inline controllers/teleport_master_controller/teleport_master_controller.cpp:107]: const these too

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (KipHamiltons) [inline controllers/teleport_master_controller/teleport_master_controller.cpp:121]: aren't there 8 rotations?

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (KipHamiltons) [inline controllers/teleport_master_controller/teleport_master_controller.cpp:146]:
```suggestion
    const webots::Node* fieldNode = supervisor.getFromDef(fieldDef);
```

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (KipHamiltons) [inline controllers/teleport_master_controller/teleport_master_controller.cpp:258]:
```suggestion
            const webots::Field* robotTranslationField = supervisor.getFromDef(selfDef)->getField("translation");
```

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (KipHamiltons) [inline controllers/teleport_master_controller/teleport_master_controller.cpp:262]:
```suggestion
            const webots::Field* robotRotationField = supervisor.getFromDef(selfDef)->getField("rotation");
```

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (KipHamiltons) [inline controllers/teleport_master_controller/teleport_master_controller.cpp:328]:
```suggestion
            const std::string countPadded = padLeft(count, 7);
```

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (KipHamiltons) [inline controllers/teleport_master_controller/teleport_master_controller.cpp:325]:
```suggestion
            const Eigen::Affine3d Hwc_l = Htw.inverse() * Htx * Hxl;
            const Eigen::Affine3d Hwc_r = Htw.inverse() * Htx * Hxr;
```

**Review by KipHamiltons — CHANGES_REQUESTED** [review]

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (ysims) [reply controllers/teleport_master_controller/teleport_master_controller.cpp:44]: This isn't padding empty characters, it's meant to pad with 0s. It's for the naming of the data files.

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (ysims) [reply controllers/teleport_master_controller/teleport_master_controller.cpp]: I've changed this to be read in from the yaml file, so I don't think it can be `constexpr` anymore.

File: controllers/teleport_master_controller/teleport_master_controller.cpp
Comment (ysims) [reply controllers/teleport_master_controller/teleport_master_controller.cpp:66]: No

File: CMakeLists.txt
Comment (ysims) [inline CMakeLists.txt:103]: This shouldn't be in this PR and needs to go before merging. Waiting on #88 to be merged since this won't work without it.

**Comment by ysims** [comment]
This is ready for a review. I've gone through it and refactored it a bit. Notable changes
- Rotations are now a random angle rather than a fixed number of preset rotations
- Heights are read from the proto file
- Main has been merged, tbh it also has the libcppcontroller stuff in it as well which is from #88 otherwise it won't run. #88 should be merged first.
- Cleaned up the code, made sure clang tidy is happy
- README didn't look like it needed updating but I expanded it a little
- I can't remember what we decided on for the goals but I'm pretty sure we want to detect the small goal posts separately and remove the net (segmentation doesn't work properly if the net is blocking it). So I've put this back in.
- Renamed the controller to something that better describes what it does

**Comment by ysims** [comment]
@LachlanCourt informal request for review lol since it's your PR so I can't request it properly :stuck_out_tongue_closed_eyes:

**Comment by LachlanCourt** [comment]
I'll have a look tomorrow 😊

**Comment by LachlanCourt** [comment]
This works for me :+1: I can't approve it tho bc it's my own PR initially :laughing:

**Review by ysims — DISMISSED** [review]
Lachlan and I think this is good so I'm approving it 😆 😅

**Review by BaileyJoe — APPROVED** [review]
Got it running seems good 👍

**Comment by ysims** [comment]
Can someone review who isn't myself or @LachlanCourt pls ty

**Review by CMurtagh-LGTM — DISMISSED** [review]

File: nuwebots/controllers/vision_teleport_controller/vision_teleport_controller.cpp
Comment (ysims) [reply nuwebots/controllers/vision_teleport_controller/vision_teleport_controller.cpp:60]: It seems okay to me? At the beginning, assume there is no collision.

Comment (CMurtagh-LGTM) [reply nuwebots/controllers/vision_teleport_controller/vision_teleport_controller.cpp:60]: If we assume there is no collision why would we check for one?

Comment (ysims) [reply nuwebots/controllers/vision_teleport_controller/vision_teleport_controller.cpp:60]: because our assumption may be wrong

Comment (ysims) [reply nuwebots/controllers/vision_teleport_controller/vision_teleport_controller.cpp:60]:
```suggestion
        // Flag for if robots have collided. Start by assuming no collision, and then update if there is any collision.
        bool collision = false;
```

**Review by CMurtagh-LGTM — APPROVED** [review]

**Pull request merged**
